lighting/raymarch/volume.glsl

```diff
@@ -1,103 +1,137 @@
-                if (densityLight>0.)
-                    Tl *= 1. - densityLight * absorption/fSamples;
-                if (Tl <= 0.01)
-                    break;
+vec3 shadowTransmittance(vec3 position, vec3 rayDirectionL, float stepSizeL) {
+    vec3 transmittanceL = vec3(1.0, 1.0, 1.0);
+    float tL = 0.0;
+
+    for (int j = 0; j < RAYMARCH_VOLUME_SAMPLES_LIGHT; j++) {                
+        vec3 positionL = position + rayDirectionL * tL;
+        #if defined(RAYMARCH_VOLUMETRIC_SHADOWS)
+            Material mat = RAYMARCH_MAP_FNC(positionL);
+            if (mat.sdf <= 0.0) {
+                return vec3(0.0, 0.0, 0.0);
             }
-            col += vec4(LIGHT_COLOR * 80. * tmp * T * Tl, 1.0);
+        #endif
+        Medium resL = RAYMARCH_VOLUME_MAP_FNC(positionL);
+        float densityL = -resL.sdf;
+        vec3 extinctionL = resL.absorption + resL.scattering;
+        transmittanceL *= exp(-densityL * extinctionL * stepSizeL);
+
+        float offset = random(position)*(stepSizeL*RAYMARCH_VOLUME_DITHER);
+        tL += stepSizeL + offset;
+    }
+
+    return transmittanceL;
+}
+
+vec3 raymarchVolume(vec3 rayOrigin, vec3 rayDirection, vec2 st, float minDist, vec3 background) {
+    vec3 scatteredLuminance = vec3(0.0, 0.0, 0.0);
+    vec3 transmittance = vec3(1.0, 1.0, 1.0);
+    float stepSize = RAYMARCH_MAX_DIST/float(RAYMARCH_VOLUME_SAMPLES);
+
+    float t = RAYMARCH_MIN_DIST;
+
+    for(int i = 0; i < RAYMARCH_VOLUME_SAMPLES; i++) {        
+        vec3 position = rayOrigin + rayDirection * t;
+        Medium res = RAYMARCH_VOLUME_MAP_FNC(position);
+        float density = -res.sdf;
+        vec3 extinction = res.absorption + res.scattering;
+
+        if (t < minDist && density > 0.0) {
+
+            #if defined(LIGHT_DIRECTION) || defined(LIGHT_POSITION)
+                #if defined(LIGHT_DIRECTION) // directional light
+                    float stepSizeL = RAYMARCH_MAX_DIST/float(RAYMARCH_VOLUME_SAMPLES_LIGHT);
+                    vec3 rayDirectionL = normalize(LIGHT_DIRECTION);
+                    const float attenuationL = 1.0;
+                #else // point light
+                    float distL = distance(LIGHT_POSITION, position);
+                    float stepSizeL = distL/float(RAYMARCH_VOLUME_SAMPLES_LIGHT);
+                    vec3 rayDirectionL = normalize(LIGHT_POSITION - position);
+                    float attenuationL = attenuation(distL);
+                #endif
+                vec3 shadow = shadowTransmittance(position, rayDirectionL, stepSizeL);
+                vec3 L = LIGHT_COLOR * LIGHT_INTENSITY;
+            #else // no lighting
+                const float attenuationL = 1.0;
+                const vec3 shadow = 1.0;
+                const vec3 L = vec3(1.0, 1.0, 1.0);
+            #endif
+
+            #if defined RAYMARCH_ENERGY_CONSERVING
+                // energy-conserving scattering integration
+                vec3 S = L * attenuationL * shadow * density * res.scattering;
```

Comment:
```
vec3 S = L * attenuationL * shadow * density * scattering;
```

Reply:
@patriciogonzalezvivo Not sure I see what's the issue here. `res.scattering`looks correct to me. `scaterring`is not defined.